Move getambassador.io/config annotation unfolding to happen in entrypoint.go (part 2; Python)

LukeShu opened the pull request:
## Description
This PR is part 2 of an update to https://github.com/emissary-ingress/emissary/pull/2271.  It completes the Python-side of the change, and removes a bunch of now-dead code.

As usual for me, I suggest a commit-by-commit review.

## Related Issues
- https://github.com/emissary-ingress/emissary/pull/2271 (original version)
- https://github.com/emissary-ingress/emissary/pull/4014 (part 1)
- https://github.com/emissary-ingress/emissary/issues/4006 (issue bringing apiVersion issues to light)
- https://github.com/emissary-ingress/emissary/pull/4029 (an issue discovered by this work)

## Testing
CI should adequately verify that this doesn't break anything. This PR doesn't add any testing, though it would benefit from the "run the v1.14 tests" work.

## Checklist

<!--
  Please review the requirements for each checkbox, and check them
  off (change "[ ]" to "[x]") as you verify that they are complete.
-->

 - [x] I made sure to update `CHANGELOG.md`. - yes
   
   Remember, the CHANGELOG needs to mention:
    + Any new features
    + Any changes to our included version of Envoy
    + Any non-backward-compatible changes
    + Any deprecations
 
 - [ ] This is unlikely to impact how Ambassador performs at scale. - actually, it should improve things
 
   Remember, things that might have an impact at scale include:
    + Any significant changes in memory use that might require adjusting the memory limits
    + Any significant changes in CPU use that might require adjusting the CPU limits
    + Anything that might change how many replicas users should use
    + Changes that impact data-plane latency/scalability
 
 - [ ] My change is adequately tested. - no, this would benefit from the "use the v1.14 tests" work
 
   Remember when considering testing:
    + Your change needs to be specifically covered by tests.
       + Tests need to cover all the states where your change is relevant: for example, if you add a behavior that can be enabled or disabled, you'll need tests that cover the enabled case and tests that cover the disabled case. It's not sufficient just to test with the behavior enabled.
    + You also need to make sure that the _entire area being changed_ has adequate test coverage.
       + If existing tests don't actually cover the entire area being changed, add tests.
       + This applies even for aspects of the area that you're not changing – check the test coverage, and improve it if needed!
    + We should lean on the bulk of code being covered by unit tests, but...
    + ... an end-to-end test should cover the integration points
 
 - [x] I updated `DEVELOPING.md` with any any special dev tricks I had to use to work on this code efficiently. - no tricks

kflynn reviewed (dismissed):
This looks great. 🙂 It makes me especially happy to see `Host.proto` and all the schemas _go away_. 😁

LukeShu commented:
v2:
```patch
diff --git a/.github/workflows/execute-tests-and-promote.yml b/.github/workflows/execute-tests-and-promote.yml
index 18b54737a..8691b148e 100644
--- a/.github/workflows/execute-tests-and-promote.yml
+++ b/.github/workflows/execute-tests-and-promote.yml
@@ -70,7 +70,19 @@ jobs:
       DOCKER_PASSWORD: ${{ secrets.GH_DOCKER_BUILD_TOKEN }}
     strategy:
       matrix:
-        test: [envoy-ah, envoy-v2-ah, envoy-ip, envoy-v2-ip, envoy-qz, envoy-v2-qz, kat, integration]
+        test:
+          - integration
+          - kat-envoy2-ah
+          - kat-envoy2-ip
+          - kat-envoy2-qz
+          - kat-envoy3-ah
+          - kat-envoy3-ip
+          - kat-envoy3-qz
+          # FIXME(lukeshu): KAT_RUN_MODE=local is disabled because it
+          # needs fixed for a world where annotations are already
+          # unfolded in the snapshot.
+          #
+          #- kat-local
     name: pytest-${{ matrix.test }}
     steps:
       - uses: actions/checkout@v2
diff --git a/builder/builder.mk b/builder/builder.mk
index fc0af1af2..45098807e 100644
--- a/builder/builder.mk
+++ b/builder/builder.mk
@@ -455,10 +455,17 @@ pytest-integration:
 	$(MAKE) pytest PYTEST_ARGS="$$PYTEST_ARGS python/tests/integration"
 .PHONY: pytest-integration
 
-pytest-kat:
-	@printf "$(CYN)==> $(GRN)Running $(BLU)py$(GRN) kat tests$(END)\n"
+pytest-kat-local:
 	$(MAKE) pytest PYTEST_ARGS="$$PYTEST_ARGS python/tests/kat"
-.PHONY: pytest-kat
+pytest-kat-envoy3: # doing this all at once is too much for CI...
+	$(MAKE) pytest KAT_RUN_MODE=envoy PYTEST_ARGS="$$PYTEST_ARGS python/tests/kat"
+pytest-kat-envoy3-%: # ... so we have a separate rule to run things split up
+	$(MAKE) pytest KAT_RUN_MODE=envoy PYTEST_ARGS="$$PYTEST_ARGS --letter-range $* python/tests/kat"
+pytest-kat-envoy2: # doing this all at once is too much for CI...
+	$(MAKE) pytest KAT_RUN_MODE=envoy AMBASSADOR_ENVOY_API_VERSION=V2 PYTEST_ARGS="$$PYTEST_ARGS python/tests/kat"
+pytest-kat-envoy2-%: # ... so we have a separate rule to run things split up
+	$(MAKE) pytest KAT_RUN_MODE=envoy AMBASSADOR_ENVOY_API_VERSION=V2 PYTEST_ARGS="$$PYTEST_ARGS --letter-range $* python/tests/kat"
+.PHONY: pytest-kat-%
 
 extract-bin-envoy: docker/base-envoy.docker.tag.local
 	@mkdir -p $(OSS_HOME)/bin/
@@ -474,52 +481,13 @@ pytest-builder: test-ready
 	$(MAKE) pytest-builder-only
 .PHONY: pytest-builder
 
-pytest-envoy-ah:
-	@printf "$(CYN)==> $(GRN)Running $(BLU)py envoy$(GRN) kat tests (ah) $(END)\n"
-	$(MAKE) pytest KAT_RUN_MODE=envoy PYTEST_ARGS="$$PYTEST_ARGS --letter-range ah python/tests/kat"
-.PHONY: pytest-envoy-ah
-
-pytest-envoy-ip:
-	@printf "$(CYN)==> $(GRN)Running $(BLU)py envoy$(GRN) kat tests (ip)$(END)\n"
-	$(MAKE) pytest KAT_RUN_MODE=envoy PYTEST_ARGS="$$PYTEST_ARGS --letter-range ip python/tests/kat"
-.PHONY: pytest-envoy-ip
-
-pytest-envoy-qz:
-	@printf "$(CYN)==> $(GRN)Running $(BLU)py envoy$(GRN) kat tests (qz)$(END)\n"
-	$(MAKE) pytest KAT_RUN_MODE=envoy PYTEST_ARGS="$$PYTEST_ARGS --letter-range qz python/tests/kat"
-.PHONY: pytest-envoy-qz
-
-
-pytest-envoy:
-	$(MAKE) pytest KAT_RUN_MODE=envoy PYTEST_ARGS="$$PYTEST_ARGS python/tests/kat"
-.PHONY: pytest-envoy
-
-pytest-envoy-builder:
+pytest-envoy3-builder:
 	$(MAKE) pytest-builder KAT_RUN_MODE=envoy
-.PHONY: pytest-envoy-builder
-
-pytest-envoy-v2-ah:
-	@printf "$(CYN)==> $(GRN)Running $(BLU)py envoy v2$(GRN) kat tests (ah)$(END)\n"
-	$(MAKE) pytest KAT_RUN_MODE=envoy AMBASSADOR_ENVOY_API_VERSION=V2 PYTEST_ARGS="$$PYTEST_ARGS --letter-range ah python/tests/kat"
-.PHONY: pytest-envoy-ah
-
-pytest-envoy-v2-ip:
-	@printf "$(CYN)==> $(GRN)Running $(BLU)py envoy v2$(GRN) kat tests (ip)$(END)\n"
-	$(MAKE) pytest KAT_RUN_MODE=envoy AMBASSADOR_ENVOY_API_VERSION=V2 PYTEST_ARGS="$$PYTEST_ARGS --letter-range ip python/tests/kat"
-.PHONY: pytest-envoy-v2-ip
-
-pytest-envoy-v2-qz:
-	@printf "$(CYN)==> $(GRN)Running $(BLU)py envoy v2$(GRN) kat tests (qz)$(END)\n"
-	$(MAKE) pytest KAT_RUN_MODE=envoy AMBASSADOR_ENVOY_API_VERSION=V2 PYTEST_ARGS="$$PYTEST_ARGS --letter-range qz python/tests/kat"
-.PHONY: pytest-envoy-v2-qz
-
-pytest-envoy-v2:
-	$(MAKE) pytest KAT_RUN_MODE=envoy AMBASSADOR_ENVOY_API_VERSION=V2 PYTEST_ARGS="$$PYTEST_ARGS python/tests/kat"
-.PHONY: pytest-envoy-v2
+.PHONY: pytest-envoy3-builder
 
-pytest-envoy-v2-builder:
+pytest-envoy2-builder:
 	$(MAKE) pytest-builder KAT_RUN_MODE=envoy AMBASSADOR_ENVOY_API_VERSION=V2
-.PHONY: pytest-envoy-v2-builder
+.PHONY: pytest-envoy2-builder
 
 pytest-builder-only: sync preflight-cluster | docker/$(LCNAME).docker.push.remote docker/kat-client.docker.push.remote docker/kat-server.docker.push.remote
 	@printf "$(CYN)==> $(GRN)Running $(BLU)py$(GRN) tests in builder shell$(END)\n"
diff --git a/python/kat/harness.py b/python/kat/harness.py
index bffca0584..26f201fbb 100755
--- a/python/kat/harness.py
+++ b/python/kat/harness.py
@@ -41,10 +41,7 @@ except AttributeError:
 
 # Run mode can be local (don't do any Envoy stuff), envoy (only do Envoy stuff),
 # or all (allow both). Default is all.
-#
-# FIXME(lukeshu): KAT_RUN_MODE=local is disabled because it needs fixed for a world where
-# annotations are already unfolded in the snapshot.
-RUN_MODE = 'envoy' # os.environ.get('KAT_RUN_MODE', 'all').lower()
+RUN_MODE = os.environ.get('KAT_RUN_MODE', 'all').lower()
 
 # We may have a SOURCE_ROOT override from the environment
 SOURCE_ROOT = os.environ.get('SOURCE_ROOT', '')
```

LukeShu commented:
v3:
- rebase on to `release/v2.1`
- fix dict iteration:
```patch
diff --git a/python/ambassador/fetch/fetcher.py b/python/ambassador/fetch/fetcher.py
index 48dadd896..84b782e69 100644
--- a/python/ambassador/fetch/fetcher.py
+++ b/python/ambassador/fetch/fetcher.py
@@ -283,7 +283,7 @@ spec:
                 watt_list.append(obj)
 
             # Now we'll do the same thing for unfolded annotations.
-            for parent_key, ann_list in watt_k8s.pop('annotations') or {}:
+            for parent_key, ann_list in (watt_k8s.pop('annotations') or {}).items():
                 for obj in ann_list.get('valid', []) + ann_list.get('invalid', []):
                     kind = obj['kind']
 
```

LukeShu commented:
v4:
```patch
diff --git a/python/ambassador/fetch/fetcher.py b/python/ambassador/fetch/fetcher.py
index 84b782e69..ed6859b66 100644
--- a/python/ambassador/fetch/fetcher.py
+++ b/python/ambassador/fetch/fetcher.py
@@ -275,7 +275,6 @@ spec:
                 # explicit null values -- they'll need to be turned into empty lists
                 # and re-saved, and setdefault() won't do that for an explicit null.
                 watt_list = watt_k8s.get(kind)
-
                 if not watt_list:
                     watt_list = []
                     watt_k8s[kind] = watt_list
@@ -288,7 +287,6 @@ spec:
                     kind = obj['kind']
 
                     watt_list = watt_k8s.get(kind)
-
                     if not watt_list:
                         watt_list = []
                         watt_k8s[kind] = watt_list
diff --git a/python/ambassador/fetch/resource.py b/python/ambassador/fetch/resource.py
index 4fc6f5830..636da6d7c 100644
--- a/python/ambassador/fetch/resource.py
+++ b/python/ambassador/fetch/resource.py
@@ -56,15 +56,21 @@ class NormalizedResource:
 
     @classmethod
     def from_kubernetes_object(cls, obj: KubernetesObject, rkey: Optional[str] = None) -> NormalizedResource:
-        if obj.gvk.api_group not in ('getambassador.io', 'getambassador.io'):
-            raise ValueError(f'Cannot construct resource from non-Ambassador Kubernetes object with API version {obj.gvk.api_version}')
         if obj.namespace is None:
             raise ValueError(f'Cannot construct resource from Kubernetes object {obj.key} without namespace')
 
         labels = dict(obj.labels)
-        if not rkey:
+
+        if not rkey:  # rkey is only set for annotations
+            # Default rkey for native Kubernetes resources
             rkey = f"{obj.name}.{obj.namespace}"
+            # Some other code uses the 'ambassador_crd' label to know which resource to update
+            # .status for with the apiserver.  Which is (IMO) a horrible hack, but I'm not up for
+            # changing it at the moment.
             labels['ambassador_crd'] = rkey
+        else:
+            # Don't let it think that an annotation can have its status updated.
+            labels.pop('ambassador_crd', None)
 
         # When creating an Ambassador object from a Kubernetes object, we have to make
         # sure that we pay attention to 'errors', which will be set IFF watt's validation
diff --git a/python/tests/kat/t_basics.py b/python/tests/kat/t_basics.py
index 89b3880fb..de81bbf49 100644
--- a/python/tests/kat/t_basics.py
+++ b/python/tests/kat/t_basics.py
@@ -122,6 +122,15 @@ spec:
   prefix_bad: /bad-<<WHICH>>/
   service: {self.target.path.fqdn}
 """, """
+kind: Mapping
+metadata:
+  name:  {self.path.k8s}-m-bad-no-apiversion-<<WHICH>>
+spec:
+  ambassador_id: ["{self.ambassador_id}"]
+  hostname: "*"
+  prefix_bad: /bad-<<WHICH>>/
+  service: {self.target.path.fqdn}
+""", """
 apiVersion: getambassador.io/v3alpha1
 kind:  Module
 metadata:
@@ -207,12 +216,13 @@ spec:
 
         for m in self.models:
             m_yaml = self.format(m.replace("<<WHICH>>", "crd"))
+            m_obj = yaml.safe_load(m_yaml)
+            if 'apiVersion' not in m_obj:
+                continue
 
             manifests.append("---")
             manifests.append(m_yaml)
 
-            m_obj = yaml.safe_load(m_yaml)
-
             if 'good' not in m_obj["metadata"]["name"]:
                 self.resource_names.append(m_obj["metadata"]["name"] + ".default.1")
 
diff --git a/python/tests/kat/t_redirect.py b/python/tests/kat/t_redirect.py
index 173b5c704..b81e16698 100644
--- a/python/tests/kat/t_redirect.py
+++ b/python/tests/kat/t_redirect.py
@@ -262,6 +262,7 @@ spec:
 
     def config(self) -> Generator[Union[str, Tuple[Node, str]], None, None]:
         yield self.target, self.format("""
+apiVersion: getambassador.io/v3alpha1
 kind: Module
 name: ambassador
 config:
```

LukeShu commented:
Changes from the last time @kflynn reviewed it:
```patch
diff --git a/.github/workflows/execute-tests-and-promote.yml b/.github/workflows/execute-tests-and-promote.yml
index a0a8602cf..4403ce215 100644
--- a/.github/workflows/execute-tests-and-promote.yml
+++ b/.github/workflows/execute-tests-and-promote.yml
@@ -71,7 +71,19 @@ jobs:
     strategy:
       fail-fast: false
       matrix:
-        test: [envoy-ah, envoy-v2-ah, envoy-ip, envoy-v2-ip, envoy-qz, envoy-v2-qz, kat, integration]
+        test:
+          - integration
+          - kat-envoy2-ah
+          - kat-envoy2-ip
+          - kat-envoy2-qz
+          - kat-envoy3-ah
+          - kat-envoy3-ip
+          - kat-envoy3-qz
+          # FIXME(lukeshu): KAT_RUN_MODE=local is disabled because it
+          # needs fixed for a world where annotations are already
+          # unfolded in the snapshot.
+          #
+          #- kat-local
     name: pytest-${{ matrix.test }}
     steps:
       - uses: actions/checkout@v2
diff --git a/builder/builder.mk b/builder/builder.mk
index e806324ef..9f51a5974 100644
--- a/builder/builder.mk
+++ b/builder/builder.mk
@@ -450,10 +450,17 @@ pytest-integration:
 	$(MAKE) pytest PYTEST_ARGS="$$PYTEST_ARGS python/tests/integration"
 .PHONY: pytest-integration
 
-pytest-kat:
-	@printf "$(CYN)==> $(GRN)Running $(BLU)py$(GRN) kat tests$(END)\n"
+pytest-kat-local:
 	$(MAKE) pytest PYTEST_ARGS="$$PYTEST_ARGS python/tests/kat"
-.PHONY: pytest-kat
+pytest-kat-envoy3: # doing this all at once is too much for CI...
+	$(MAKE) pytest KAT_RUN_MODE=envoy PYTEST_ARGS="$$PYTEST_ARGS python/tests/kat"
+pytest-kat-envoy3-%: # ... so we have a separate rule to run things split up
+	$(MAKE) pytest KAT_RUN_MODE=envoy PYTEST_ARGS="$$PYTEST_ARGS --letter-range $* python/tests/kat"
+pytest-kat-envoy2: # doing this all at once is too much for CI...
+	$(MAKE) pytest KAT_RUN_MODE=envoy AMBASSADOR_ENVOY_API_VERSION=V2 PYTEST_ARGS="$$PYTEST_ARGS python/tests/kat"
+pytest-kat-envoy2-%: # ... so we have a separate rule to run things split up
+	$(MAKE) pytest KAT_RUN_MODE=envoy AMBASSADOR_ENVOY_API_VERSION=V2 PYTEST_ARGS="$$PYTEST_ARGS --letter-range $* python/tests/kat"
+.PHONY: pytest-kat-%
 
 extract-bin-envoy: docker/base-envoy.docker.tag.local
 	@mkdir -p $(OSS_HOME)/bin/
@@ -469,52 +476,13 @@ pytest-builder: test-ready
 	$(MAKE) pytest-builder-only
 .PHONY: pytest-builder
 
-pytest-envoy-ah:
-	@printf "$(CYN)==> $(GRN)Running $(BLU)py envoy$(GRN) kat tests (ah) $(END)\n"
-	$(MAKE) pytest KAT_RUN_MODE=envoy PYTEST_ARGS="$$PYTEST_ARGS --letter-range ah python/tests/kat"
-.PHONY: pytest-envoy-ah
-
-pytest-envoy-ip:
-	@printf "$(CYN)==> $(GRN)Running $(BLU)py envoy$(GRN) kat tests (ip)$(END)\n"
-	$(MAKE) pytest KAT_RUN_MODE=envoy PYTEST_ARGS="$$PYTEST_ARGS --letter-range ip python/tests/kat"
-.PHONY: pytest-envoy-ip
-
-pytest-envoy-qz:
-	@printf "$(CYN)==> $(GRN)Running $(BLU)py envoy$(GRN) kat tests (qz)$(END)\n"
-	$(MAKE) pytest KAT_RUN_MODE=envoy PYTEST_ARGS="$$PYTEST_ARGS --letter-range qz python/tests/kat"
-.PHONY: pytest-envoy-qz
-
-
-pytest-envoy:
-	$(MAKE) pytest KAT_RUN_MODE=envoy PYTEST_ARGS="$$PYTEST_ARGS python/tests/kat"
-.PHONY: pytest-envoy
-
-pytest-envoy-builder:
+pytest-envoy3-builder:
 	$(MAKE) pytest-builder KAT_RUN_MODE=envoy
-.PHONY: pytest-envoy-builder
-
-pytest-envoy-v2-ah:
-	@printf "$(CYN)==> $(GRN)Running $(BLU)py envoy v2$(GRN) kat tests (ah)$(END)\n"
-	$(MAKE) pytest KAT_RUN_MODE=envoy AMBASSADOR_ENVOY_API_VERSION=V2 PYTEST_ARGS="$$PYTEST_ARGS --letter-range ah python/tests/kat"
-.PHONY: pytest-envoy-ah
-
-pytest-envoy-v2-ip:
-	@printf "$(CYN)==> $(GRN)Running $(BLU)py envoy v2$(GRN) kat tests (ip)$(END)\n"
-	$(MAKE) pytest KAT_RUN_MODE=envoy AMBASSADOR_ENVOY_API_VERSION=V2 PYTEST_ARGS="$$PYTEST_ARGS --letter-range ip python/tests/kat"
-.PHONY: pytest-envoy-v2-ip
-
-pytest-envoy-v2-qz:
-	@printf "$(CYN)==> $(GRN)Running $(BLU)py envoy v2$(GRN) kat tests (qz)$(END)\n"
-	$(MAKE) pytest KAT_RUN_MODE=envoy AMBASSADOR_ENVOY_API_VERSION=V2 PYTEST_ARGS="$$PYTEST_ARGS --letter-range qz python/tests/kat"
-.PHONY: pytest-envoy-v2-qz
-
-pytest-envoy-v2:
-	$(MAKE) pytest KAT_RUN_MODE=envoy AMBASSADOR_ENVOY_API_VERSION=V2 PYTEST_ARGS="$$PYTEST_ARGS python/tests/kat"
-.PHONY: pytest-envoy-v2
+.PHONY: pytest-envoy3-builder
 
-pytest-envoy-v2-builder:
+pytest-envoy2-builder:
 	$(MAKE) pytest-builder KAT_RUN_MODE=envoy AMBASSADOR_ENVOY_API_VERSION=V2
-.PHONY: pytest-envoy-v2-builder
+.PHONY: pytest-envoy2-builder
 
 pytest-builder-only: sync preflight-cluster | docker/$(LCNAME).docker.push.remote docker/kat-client.docker.push.remote docker/kat-server.docker.push.remote
 	@printf "$(CYN)==> $(GRN)Running $(BLU)py$(GRN) tests in builder shell$(END)\n"
diff --git a/cmd/entrypoint/consul.go b/cmd/entrypoint/consul.go
index f25ed3896..16b13e365 100644
--- a/cmd/entrypoint/consul.go
+++ b/cmd/entrypoint/consul.go
@@ -23,7 +23,7 @@ type consulMapping struct {
 func ReconcileConsul(ctx context.Context, consul *consul, s *snapshotTypes.KubernetesSnapshot) error {
 	var mappings []consulMapping
 	for _, list := range s.Annotations {
-		for _, a := range list.Valid {
+		for _, a := range list {
 			switch m := a.(type) {
 			case *amb.Mapping:
 				if include(m.Spec.AmbassadorID) {
diff --git a/cmd/entrypoint/endpoints.go b/cmd/entrypoint/endpoints.go
index 7fa6b1c86..68c3f0d4e 100644
--- a/cmd/entrypoint/endpoints.go
+++ b/cmd/entrypoint/endpoints.go
@@ -74,7 +74,10 @@ func (eri *endpointRoutingInfo) reconcileEndpointWatches(ctx context.Context, s
 	// includes Modules and Resolvers. When we are done with Phase one we have processed enough
 	// resources to correctly interpret Mappings.
 	for _, list := range s.Annotations {
-		for _, a := range list.Valid {
+		for _, a := range list {
+			if _, isInvalid := a.(*kates.Unstructured); isInvalid {
+				continue
+			}
 			if include(GetAmbId(ctx, a)) {
 				eri.checkResourcePhase1(ctx, a, "annotation")
 			}
@@ -121,7 +124,10 @@ func (eri *endpointRoutingInfo) reconcileEndpointWatches(ctx context.Context, s
 	}
 
 	for _, list := range s.Annotations {
-		for _, a := range list.Valid {
+		for _, a := range list {
+			if _, isInvalid := a.(*kates.Unstructured); isInvalid {
+				continue
+			}
 			if include(GetAmbId(ctx, a)) {
 				eri.checkResourcePhase2(ctx, a, "annotation")
 			}
diff --git a/cmd/entrypoint/secrets.go b/cmd/entrypoint/secrets.go
index b386c5aeb..93d8f8626 100644
--- a/cmd/entrypoint/secrets.go
+++ b/cmd/entrypoint/secrets.go
@@ -26,7 +26,10 @@ func ReconcileSecrets(ctx context.Context, s *snapshotTypes.KubernetesSnapshot)
 	// here).
 
 	for _, list := range s.Annotations {
-		for _, a := range list.Valid {
+		for _, a := range list {
+			if _, isInvalid := a.(*kates.Unstructured); isInvalid {
+				continue
+			}
 			if include(GetAmbId(ctx, a)) {
 				resources = append(resources, a)
 			}
diff --git a/pkg/snapshot/v1/annotations.go b/pkg/snapshot/v1/annotations.go
index 49f1da67b..38b5e0aa5 100644
--- a/pkg/snapshot/v1/annotations.go
+++ b/pkg/snapshot/v1/annotations.go
@@ -35,23 +35,23 @@ func (s *KubernetesSnapshot) PopulateAnnotations(ctx context.Context) error {
 	s.Annotations = make(map[string]AnnotationList)
 	var errs derror.MultiError
 	for _, r := range annotatable {
-		var annotations AnnotationList
 		key := annotationKey(r)
 		objs, err := ParseAnnotationResources(r)
 		if err != nil {
 			errs = append(errs, fmt.Errorf("%s: %w", key, err))
 			continue
 		}
-		for _, untypedObj := range objs {
+		annotations := make(AnnotationList, len(objs))
+		for i, untypedObj := range objs {
 			typedObj, err := ValidateAndConvertObject(ctx, untypedObj)
 			if err != nil {
 				untypedObj.Object["errors"] = err.Error()
-				annotations.Invalid = append(annotations.Invalid, untypedObj)
+				annotations[i] = untypedObj
 			} else {
-				annotations.Valid = append(annotations.Valid, typedObj)
+				annotations[i] = typedObj
 			}
 		}
-		if annotations.Valid != nil || annotations.Invalid != nil {
+		if len(annotations) > 0 {
 			s.Annotations[key] = annotations
 		}
 	}
diff --git a/pkg/snapshot/v1/annotations_test.go b/pkg/snapshot/v1/annotations_test.go
index 7278b121d..c90fa7ebe 100644
--- a/pkg/snapshot/v1/annotations_test.go
+++ b/pkg/snapshot/v1/annotations_test.go
@@ -125,65 +125,59 @@ prefix: /blah/`
 	assert.NoError(t, err)
 	assert.Equal(t, map[string]snapshotTypes.AnnotationList{
 		"Service/svc.ambassador": {
-			Valid: []kates.Object{
-				&amb.Mapping{
-					TypeMeta: metav1.TypeMeta{
-						Kind:       "Mapping",
-						APIVersion: "getambassador.io/v3alpha1",
-					},
-					ObjectMeta: metav1.ObjectMeta{
-						Name:      "quote-backend",
-						Namespace: "ambassador",
-					},
-					Spec: amb.MappingSpec{
-						Prefix:  "/backend/",
-						Service: "quote:80",
-					},
+			&amb.Mapping{
+				TypeMeta: metav1.TypeMeta{
+					Kind:       "Mapping",
+					APIVersion: "getambassador.io/v3alpha1",
+				},
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      "quote-backend",
+					Namespace: "ambassador",
+				},
+				Spec: amb.MappingSpec{
+					Prefix:  "/backend/",
+					Service: "quote:80",
 				},
 			},
 		},
 		"Ingress/ingress.somens": {
-			Invalid: []*kates.Unstructured{
-				{
-					Object: map[string]interface{}{
-						"apiVersion": "getambassador.io/v3alpha1",
-						"kind":       "Mapping",
-						"metadata": map[string]interface{}{
-							"name":      "cool-mapping",
-							"namespace": "somens",
-						},
-						"spec": map[string]interface{}{
-							"prefix": "/blah/",
-						},
-						"errors": "spec.service in body is required",
+			&kates.Unstructured{
+				Object: map[string]interface{}{
+					"apiVersion": "getambassador.io/v3alpha1",
+					"kind":       "Mapping",
+					"metadata": map[string]interface{}{
+						"name":      "cool-mapping",
+						"namespace": "somens",
+					},
+					"spec": map[string]interface{}{
+						"prefix": "/blah/",
 					},
+					"errors": "spec.service in body is required",
 				},
 			},
 		},
 		"Service/ambassador.ambassador": {
-			Valid: []kates.Object{
-				&amb.Module{
-					TypeMeta: metav1.TypeMeta{
-						Kind:       "Module",
-						APIVersion: "getambassador.io/v3alpha1",
-					},
-					ObjectMeta: metav1.ObjectMeta{
-						Name:      "ambassador",
-						Namespace: "ambassador",
-					},
-					Spec: amb.ModuleSpec{
-						Config: getModuleSpec(t, `{"diagnostics":{"enabled":true}}`),
-					},
+			&amb.Module{
+				TypeMeta: metav1.TypeMeta{
+					Kind:       "Module",
+					APIVersion: "getambassador.io/v3alpha1",
 				},
-				&amb.KubernetesEndpointResolver{
-					TypeMeta: metav1.TypeMeta{
-						Kind:       "KubernetesEndpointResolver",
-						APIVersion: "getambassador.io/v3alpha1",
-					},
-					ObjectMeta: metav1.ObjectMeta{
-						Name:      "endpoint",
-						Namespace: "ambassador",
-					},
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      "ambassador",
+					Namespace: "ambassador",
+				},
+				Spec: amb.ModuleSpec{
+					Config: getModuleSpec(t, `{"diagnostics":{"enabled":true}}`),
+				},
+			},
+			&amb.KubernetesEndpointResolver{
+				TypeMeta: metav1.TypeMeta{
+					Kind:       "KubernetesEndpointResolver",
+					APIVersion: "getambassador.io/v3alpha1",
+				},
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      "endpoint",
+					Namespace: "ambassador",
 				},
 			},
 		},
diff --git a/pkg/snapshot/v1/types.go b/pkg/snapshot/v1/types.go
index f267ba31b..648f45fd2 100644
--- a/pkg/snapshot/v1/types.go
+++ b/pkg/snapshot/v1/types.go
@@ -92,7 +92,7 @@ type KubernetesSnapshot struct {
 	FSSecrets  map[SecretRef]*kates.Secret `json:"-"`      // Secrets from the filesystem
 	Secrets    []*kates.Secret             `json:"secret"` // Secrets we'll feed to Ambassador
 
-	// [kind/name.namespace]AnnotationList
+	// [kind/name.namespace][]kates.Object
 	Annotations map[string]AnnotationList `json:"annotations"`
 
 	// Pods, Deployments and ConfigMaps were added to be used by Ambassador Agent so it can
@@ -116,10 +116,12 @@ type KubernetesSnapshot struct {
 	ArgoApplications []*kates.Unstructured `json:"ArgoApplications,omitempty"`
 }
 
-type AnnotationList struct {
-	Valid   []kates.Object        `json:"valid,omitempty"`
-	Invalid []*kates.Unstructured `json:"invalid,omitempty"`
-}
+// AnnotationList is a []kates.Object that round-trips through JSON (kates.Object is an interface,
+// and you can't normally unmarshal in to an interface).
+//
+// The kates.Object will be the appropriate struct(-pointer) type for valid resources, and a
+// *kates.Unstructured for invalid resources.
+type AnnotationList []kates.Object
 
 // UnmarshalJSON implements json.Unmarshaler, and exists because unmarshalling directly in to an
 // interface (kates.Object) doesn't work.
@@ -128,28 +130,28 @@ func (al *AnnotationList) UnmarshalJSON(bs []byte) error {
 		*al = AnnotationList{}
 		return nil
 	}
-	var ul struct {
-		Valid   []*kates.Unstructured `json:"valid,omitempty"`
-		Invalid []*kates.Unstructured `json:"invalid,omitempty"`
-	}
+	var untyped []*kates.Unstructured
 
 	// Unmarshal as unstructured
-	err := json.Unmarshal(bs, &ul)
-	al.Invalid = ul.Invalid
+	err := json.Unmarshal(bs, &untyped)
 	if err != nil {
 		return err
 	}
 
-	// Convert the unstructured ul.Valid to the typed al.Valid.
-	al.Valid = make([]kates.Object, 0, len(ul.Valid))
-	for _, inObj := range ul.Valid {
-		outObj, err := convertAnnotationObject(inObj)
-		if err != nil {
-			return err
+	typed := make(AnnotationList, len(untyped))
+	for i, inObj := range untyped {
+		if _, isInvalid := inObj.Object["errors"]; isInvalid {
+			typed[i] = inObj
+		} else {
+			outObj, err := convertAnnotationObject(inObj)
+			if err != nil {
+				return err
+			}
+			typed[i] = outObj
 		}
-		al.Valid = append(al.Valid, outObj)
 	}
 
+	*al = typed
 	return nil
 }
 
diff --git a/python/ambassador/fetch/fetcher.py b/python/ambassador/fetch/fetcher.py
index 48dadd896..7ec050f0f 100644
--- a/python/ambassador/fetch/fetcher.py
+++ b/python/ambassador/fetch/fetcher.py
@@ -210,7 +210,8 @@ spec:
                 # UGH. This parse_yaml is the one we imported from utils. XXX This needs to be fixed.
                 for obj in parse_yaml(serialization):
                     if k8s:
-                        self.handle_k8s(obj)
+                        with self.manager.locations.push_reset():
+                            self.handle_k8s(obj)
                     else:
                         self.manager.emit(NormalizedResource(obj, rkey=rkey))
             except yaml.error.YAMLError as e:
@@ -268,8 +269,16 @@ spec:
             # Why, you may ask, do we want to dump invalid objects back in to be
             # processed??? It's because they have error information that we need to
             # propagate to the user, and this is the simplest way to do that.
-            for obj in watt_dict.get('Invalid') or []:
-                kind = obj['kind']
+
+            invalid: List[Dict] = watt_dict.get('Invalid') or []
+
+            for obj in invalid:
+                kind = obj.get('kind', None)
+
+                if not kind:
+                    # Can't work with this at _all_.
+                    self.logger.error(f"skipping invalid object with no kind: {obj}")
+                    continue
 
                 # We can't use watt_k8s.setdefault() here because many keys have
                 # explicit null values -- they'll need to be turned into empty lists
@@ -282,19 +291,8 @@ spec:
 
                 watt_list.append(obj)
 
-            # Now we'll do the same thing for unfolded annotations.
-            for parent_key, ann_list in watt_k8s.pop('annotations') or {}:
-                for obj in ann_list.get('valid', []) + ann_list.get('invalid', []):
-                    kind = obj['kind']
-
-                    watt_list = watt_k8s.get(kind)
-
-                    if not watt_list:
-                        watt_list = []
-                        watt_k8s[kind] = watt_list
-
-                    obj['_parent_key'] = parent_key
-                    watt_list.append(obj)
+            # Remove annotations from the snapshot; we'll process them separately.
+            annotations = watt_k8s.pop('annotations') or {}
 
             # These objects have to be processed first, in order, as they depend
             # on each other.
@@ -309,11 +307,12 @@ spec:
             for key in dict.fromkeys(watt_k8s_keys):
                 for obj in watt_k8s.get(key) or []:
                     # self.logger.debug(f"Handling Kubernetes {key}...")
-                    parent_key = obj.pop('_parent_key', None)
-                    if parent_key:
-                        self.handle_annotation(parent_key, obj)
-                    else:
+                    with self.manager.locations.push_reset():
                         self.handle_k8s(obj)
+                        if 'errors' not in obj:
+                            ann_parent_key = f"{obj['kind']}/{obj['metadata']['name']}.{obj['metadata'].get('namespace')}"
+                            for ann_obj in (annotations.get(ann_parent_key) or []):
+                                self.handle_annotation(ann_parent_key, ann_obj)
 
             watt_consul = watt_dict.get('Consul', {})
             consul_endpoints = watt_consul.get('Endpoints', {})
@@ -360,9 +359,8 @@ spec:
             # can't process it.
             return
 
-        with self.manager.locations.push_reset():
-            if not self.k8s_processor.try_process(obj):
-                self.logger.debug(f"{self.location}: skipping K8s {obj.gvk}")
+        if not self.k8s_processor.try_process(obj):
+            self.logger.debug(f"{self.location}: skipping K8s {obj.gvk}")
 
     def handle_annotation(self, parent_key: str, raw_obj: dict) -> None:
         try:
@@ -372,10 +370,9 @@ spec:
             # can't process it.
             return
 
-        with self.manager.locations.push_reset():
-            with self.manager.locations.mark_annotated():
-                rkey = parent_key.split('/', 1)[1]
-                self.manager.emit(NormalizedResource.from_kubernetes_object(obj, rkey=rkey))
+        with self.manager.locations.mark_annotated():
+            rkey = parent_key.split('/', 1)[1]
+            self.manager.emit(NormalizedResource.from_kubernetes_object(obj, rkey=rkey))
 
     # Handler for Consul services
     def handle_consul_service(self,
diff --git a/python/ambassador/fetch/resource.py b/python/ambassador/fetch/resource.py
index 4fc6f5830..636da6d7c 100644
--- a/python/ambassador/fetch/resource.py
+++ b/python/ambassador/fetch/resource.py
@@ -56,15 +56,21 @@ class NormalizedResource:
 
     @classmethod
     def from_kubernetes_object(cls, obj: KubernetesObject, rkey: Optional[str] = None) -> NormalizedResource:
-        if obj.gvk.api_group not in ('getambassador.io', 'getambassador.io'):
-            raise ValueError(f'Cannot construct resource from non-Ambassador Kubernetes object with API version {obj.gvk.api_version}')
         if obj.namespace is None:
             raise ValueError(f'Cannot construct resource from Kubernetes object {obj.key} without namespace')
 
         labels = dict(obj.labels)
-        if not rkey:
+
+        if not rkey:  # rkey is only set for annotations
+            # Default rkey for native Kubernetes resources
             rkey = f"{obj.name}.{obj.namespace}"
+            # Some other code uses the 'ambassador_crd' label to know which resource to update
+            # .status for with the apiserver.  Which is (IMO) a horrible hack, but I'm not up for
+            # changing it at the moment.
             labels['ambassador_crd'] = rkey
+        else:
+            # Don't let it think that an annotation can have its status updated.
+            labels.pop('ambassador_crd', None)
 
         # When creating an Ambassador object from a Kubernetes object, we have to make
         # sure that we pay attention to 'errors', which will be set IFF watt's validation
diff --git a/python/ambassador/ir/ircors.py b/python/ambassador/ir/ircors.py
index b25621916..06ee78395 100644
--- a/python/ambassador/ir/ircors.py
+++ b/python/ambassador/ir/ircors.py
@@ -36,19 +36,10 @@ class IRCORS (IRResource):
             if value:
                 new_kwargs[to_key] = self._cors_normalize(value)
 
-        # 'origins' cannot be treated like other keys, because if it's a
-        # list, then it remains as is, but if it's a string, then it's
-        # converted to a list.  It has already been validated by the fetcher
-        # to either be a string or a list of strings.
-        #
-        # (In fact in v3alpha1 it _must_ be a list of strings, but if the
-        # resource was in an annotation, it might be a string.)
+        # 'origins' cannot be treated like other keys, because we have to transform it; Envoy wants
+        # it in a different shape than it is in the CRD.
         origins = kwargs.get('origins', None)
-
         if origins is not None:
-            if type(origins) is not list:
-                origins = origins.split(',')
-
             new_kwargs['allow_origin_string_match'] = [{'exact': origin} for origin in origins]
 
         super().__init__(
diff --git a/python/kat/harness.py b/python/kat/harness.py
index bffca0584..26f201fbb 100755
--- a/python/kat/harness.py
+++ b/python/kat/harness.py
@@ -41,10 +41,7 @@ except AttributeError:
 
 # Run mode can be local (don't do any Envoy stuff), envoy (only do Envoy stuff),
 # or all (allow both). Default is all.
-#
-# FIXME(lukeshu): KAT_RUN_MODE=local is disabled because it needs fixed for a world where
-# annotations are already unfolded in the snapshot.
-RUN_MODE = 'envoy' # os.environ.get('KAT_RUN_MODE', 'all').lower()
+RUN_MODE = os.environ.get('KAT_RUN_MODE', 'all').lower()
 
 # We may have a SOURCE_ROOT override from the environment
 SOURCE_ROOT = os.environ.get('SOURCE_ROOT', '')
diff --git a/python/tests/kat/t_basics.py b/python/tests/kat/t_basics.py
index 89b3880fb..2c23354e0 100644
--- a/python/tests/kat/t_basics.py
+++ b/python/tests/kat/t_basics.py
@@ -1,3 +1,4 @@
+import re
 from typing import Generator, Tuple, Union
 
 import yaml
@@ -122,6 +123,15 @@ spec:
   prefix_bad: /bad-<<WHICH>>/
   service: {self.target.path.fqdn}
 """, """
+kind: Mapping
+metadata:
+  name:  {self.path.k8s}-m-bad-no-apiversion-<<WHICH>>
+spec:
+  ambassador_id: ["{self.ambassador_id}"]
+  hostname: "*"
+  prefix_bad: /bad-<<WHICH>>/
+  service: {self.target.path.fqdn}
+""", """
 apiVersion: getambassador.io/v3alpha1
 kind:  Module
 metadata:
@@ -207,12 +217,13 @@ spec:
 
         for m in self.models:
             m_yaml = self.format(m.replace("<<WHICH>>", "crd"))
+            m_obj = yaml.safe_load(m_yaml)
+            if 'apiVersion' not in m_obj:
+                continue
 
             manifests.append("---")
             manifests.append(m_yaml)
 
-            m_obj = yaml.safe_load(m_yaml)
-
             if 'good' not in m_obj["metadata"]["name"]:
                 self.resource_names.append(m_obj["metadata"]["name"] + ".default.1")
 
@@ -235,26 +246,20 @@ spec:
         error_dict = {}
 
         for resource, error in errors:
-            error_dict[resource] = error.split("\n", 1)[0]
+            error_dict[resource] = error
 
         for name in self.resource_names:
             assert name in error_dict, f"no error found for {name}"
 
             error = error_dict[name]
 
-            # This is a little weird. The way fast-reconfigure works with the Golang
-            # stuff, the empty config we pass in our bad Module turns into None. Python
-            # validation still catches it, but the error message is different.
-
-            # Don't be too picky about the serialization
-            expected_error = [
-                # if .spec.config is omitted because it's empty
-                "not a valid Module: None is not of type 'object'",
-                # if .spec.config is present-but-null
-                'spec.config in body must be of type object: "null"'
+            # Check that the error is one of the errors that we expect.
+            expected_errors = [
+                re.compile(r'^.* in body is required$'),
+                re.compile(r'^apiVersion None/ unsupported$'),
+                re.compile(r'^spec\.config in body must be of type object: "null"$'),
             ]
-            if error not in expected_error:
-                assert 'required' in error, f"error for {name} should talk about required properties: {error}"
+            assert any(pat.match(error) for pat in expected_errors), f"error for {name} should match one of the expected errors: {repr(error)}"
 
 
 class ServerNameTest(AmbassadorTest):
diff --git a/python/tests/kat/t_error_response.py b/python/tests/kat/t_error_response.py
index ab514b556..6163c9bef 100644
--- a/python/tests/kat/t_error_response.py
+++ b/python/tests/kat/t_error_response.py
@@ -867,7 +867,7 @@ error_response_overrides:
 - on_status_code: 503
   body:
     json_format:
-      y: "2"
+      "y": "2"
       status: '%RESPONSE_CODE%'
 '''
 
diff --git a/python/tests/kat/t_redirect.py b/python/tests/kat/t_redirect.py
index 173b5c704..b81e16698 100644
--- a/python/tests/kat/t_redirect.py
+++ b/python/tests/kat/t_redirect.py
@@ -262,6 +262,7 @@ spec:
 
     def config(self) -> Generator[Union[str, Tuple[Node, str]], None, None]:
         yield self.target, self.format("""
+apiVersion: getambassador.io/v3alpha1
 kind: Module
 name: ambassador
 config:
```

kflynn reviewed (commented):
For the record, this looks good to me. 🙂
